best-practices/rspec.md

```diff
@@ -0,0 +1,30 @@
+## Rspec style guide
+
+- First #describe what you are doing
+- Then establish the #context
+- #it only expects one thing (within reason)
+- Prefer explicitness - #it #its #specify #subject are nice shortcuts but
+consider whether they are sacrificing readability
+- Having excessive setup (fixtures, database entities, doubles etc.) can cause false-positives. Thus, set up preconditions as close to the test as possible (usually within the context block) and tear them down afterwards. This is really important when working with time for example.
+- Unit tests are fast and effective to work with when doing TDD. These are isolated tests, typically
+for a single class, and should use stubs and spies via test doubles to replace anything that is
+external to the _unit_ being tested. Testing side-effects that occur elsewhere in the system
+is incorrect. Not being able to unit test something is typically a signal that something may be wrong
+with the design of the unit
```

Comment:
I like JBRains idea of verifying contracts all the way down.

Even if we can't do this automatically with bogus or rspec-contracts (awww, :cry:) maybe we should make it a rule that any time you add or edit a stub on one unit you verify that that stub (including inputs and outputs) is tested on all lower level units that might receive that call.

So in @brentonannan's example you'd stub all the workflow's calls to services including params and returns and then in the specs on those services you'd have matching specs that passed in those params and expected those returns.